cli/server.go

```diff
@@ -2167,6 +2132,90 @@ func ConfigureHTTPServers(inv *clibase.Invocation, cfg *codersdk.DeploymentValue
 	return httpServers, nil
 }
 
+func createDERPMap(ctx context.Context, logger slog.Logger, vals *codersdk.DeploymentValues) (*tailcfg.DERPMap, error) {
+	_, accessURLPortRaw, _ := net.SplitHostPort(vals.AccessURL.Host)
+	if accessURLPortRaw == "" {
+		accessURLPortRaw = "80"
+		if vals.AccessURL.Scheme == "https" {
+			accessURLPortRaw = "443"
+		}
+	}
+
+	accessURLPort, err := strconv.Atoi(accessURLPortRaw)
+	if err != nil {
+		return nil, xerrors.Errorf("parse access URL port: %w", err)
+	}
+
+	defaultRegion := &tailcfg.DERPRegion{
+		EmbeddedRelay: true,
+		RegionID:      int(vals.DERP.Server.RegionID.Value()),
+		RegionCode:    vals.DERP.Server.RegionCode.String(),
+		RegionName:    vals.DERP.Server.RegionName.String(),
+		Nodes: []*tailcfg.DERPNode{{
+			Name:      fmt.Sprintf("%db", vals.DERP.Server.RegionID),
+			RegionID:  int(vals.DERP.Server.RegionID.Value()),
+			HostName:  vals.AccessURL.Value().Hostname(),
+			DERPPort:  accessURLPort,
+			STUNPort:  -1,
+			ForceHTTP: vals.AccessURL.Scheme == "http",
+		}},
+	}
+	if !vals.DERP.Server.Enable {
+		defaultRegion = nil
+	}
+
+	baseMapURL := vals.DERP.Config.URL.String()
+	if vals.DERP.Config.Path.String() != "" {
+		if baseMapURL != "" {
+			return nil, xerrors.Errorf("cannot specify both --derp-config-url and --derp-config-path")
+		}
+		baseMapURL = "file:" + vals.DERP.Config.Path.String()
+	}
+
+	// If they have a URL or path already set above, it takes precedence over
+	// the Tailscale DERP map value since Tailscale defaults to true.
+	if vals.DERP.Config.UseTailscaleDERPs.Value() {
+		if baseMapURL == "" {
+			baseMapURL = codersdk.TailscaleDERPMapURL
+
+			// If STUN is disabled on Tailscale DERPs warn about it. This is
+			// most likely a configuration mistake.
+			if vals.DERP.Config.BlockDirect.Value() {
+				logger.Warn(ctx, "the deployment has CODER_BLOCK_DIRECT set to `false`, and CODER_USE_TAILSCALE_DERPS is set to the default value `true`. Tailscale DERPs will not use STUN and direct connections will be blocked")
+			}
+		} else if baseMapURL != codersdk.TailscaleDERPMapURL {
+			// We add the check to the if statement in case they're manually
+			// setting the URL to the Tailscale DERP URL.
+			logger.Warn(ctx, "the deployment has CODER_DERP_CONFIG_URL or CODER_DERP_CONFIG_PATH set, and CODER_USE_TAILSCALE_DERPS is set to the default value `true`. Tailscale DERPs will NOT be used. Please set CODER_USE_TAILSCALE_DERPS to `false`")
+		}
+	}
+
+	// If the built-in DERP server is enabled, we disable Tailscale DERPs. If
+	// the custom map URL is not the Tailscale DERP map URL, we fail.
+	if baseMapURL == codersdk.TailscaleDERPMapURL && defaultRegion != nil {
+		logger.Warn(ctx, "the deployment has CODER_DERP_SERVER_ENABLE set to `true`, and is using CODER_USE_TAILSCALE_DERPS=true. Tailscale DERPs will NOT be used. Please set CODER_USE_TAILSCALE_DERPS=false or CODER_DERP_SERVER_ENABLE=false")
```

Comment:
I don't necessarily see the point of making them mutually exclusive. There could definitely be benefits of running them simultaneously.

Reply:
We discussed briefly on a NETGRU call, but the concern I have is that tailscale basically assumes that both sides of a connection can reach the home DERPs of each peer.  

If you're running the embedded DERP, and are behind a proxy, then it might be the case that the workspace agents can reach the embedded DERP without the proxy, but clients use the proxy, and the proxy breaks DERP.  In that case it's very likely that the agents are close to the embedded DERP and choose it as their home, and clients choose a tailscale DERP region as their home.

I think we would be able to figure it out from the logs (and thus paying customers would), but it would be tricky for users to diagnose this without our help.

------

That said, it's probably no worse than only running the embedded DERP, so I'm not 100% against running both if the end user intent is clear.  But, I think the default with no customization should be tailscale DERPs only.

If the end user has explicitly enabled the embedded DERP (or explicitly set the DERP map), and has _not_ explicitly enabled the tailscale DERPs, then they should get only the embedded DERP (or explicit DERP map), and not tailscale to minimize the likelihood that big corporate customers inadvertently send traffic to tailscale DERPs.